OSSM-8190: Add product documentation for Multi-Cluster Topology (#84640)
openshift/openshift-docs: modules/ossm-creating-certificates-for-multi-cluster-mesh.adoc

:_mod-docs-content-type: Procedure
[id="ossm-creating-certificates-for-a-multi-cluster-mesh_{context}"]
= Creating certificates for a multi-cluster mesh 

> This procedure explains how to create root and intermediate certificates for two clusters, named East and West.

Reviewer: ```suggestion
This procedure explains how to create the root and intermediate CA certificates for two clusters, named East and West.
```

Reviewer: Also, "East" and "West" aren't the names of the two clusters, but simply a way to distinguish between two clusters in an east-west configuration.

Author: done

[NOTE]
====
In this procedure, `cluster1` is the East cluster and `cluster2` is the West cluster. 

> ====

Reviewer: This note isn't needed, because the terms "cluster1" or "cluster2" never appear in this document.

Author: done

.Prerequisites

* You have OpenSSL installed locally.

.Procedure

> . Create a root certificate file named `root-ca.conf`:

Author: Done

+
.Sample root certificate file
[source,terminal]
----
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
[ req_dn ]
O = Istio
CN = Root CA
----

. Create a key for the root certificate by running the following command:
+
[source,terminal]
----

> $ openssl genrsa -out root-key.pem 4096

Reviewer: Do commands need to start with `$ `? This makes the copy function (at the top right of each code snippet) less useful, because the user then needs to remove the $ sign manually.

Author: Yes, all command must start with the $ prompt.

Reviewer: Then you'll have to add the prompt to a bunch of other commands in this PR. A lot of them don't have the prompt.

Also, you might want to suggest to either remove the copy icon from the snippets or change it so that it automatically removes the `$` sign when copying.

Author: published doc does not have this problem

----

. Create the certificate signing request by running the following command:
+
[source,terminal]
----
openssl req -sha256 -new -key root-key.pem \
  -config root-ca.conf \
  -out root-cert.csr
----

. Create a shared root certificate by running the following command:
+
[source,terminal]
----
openssl x509 -req -sha256 -days 3650 \
  -signkey root-key.pem \
  -extensions req_ext -extfile root-ca.conf \
  -in root-cert.csr \
  -out root-cert.pem
----

> 

Author: we discussed this new format and PR is updated.

> . Create an intermediate certificate for the East cluster by running the following command:

Author: done

+
[source,terminal]
----

> $ openssl genrsa -out east/ca-key.pem 4096

Reviewer: This command fails because the `east/` directory doesn't exist. You have to add a step before this one where the user creates the directory by running `mkdir east`.

Author: done

----

> . Create an OpenSSL configuration file named `intermediate.conf` for the intermediate certificate of the East cluster by copying the following sample file and saving it locally:

Reviewer: Since the west cluster will also have a file called `intermediate.conf`, you should mention the directory in which to create this file:
```suggestion
. Create an OpenSSL configuration file named `intermediate.conf` in the `east/` directory for the intermediate certificate of the East cluster by copying the following sample file and saving it locally:
```

Author: done

+
.Sample configuration file
[source,subs="attributes,verbatim"]
----

> east/intermediate.conf <1>

Author: done

[ req ]
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true, pathlen:0
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
subjectAltName=@san
[ san ]
DNS.1 = istiod.istio-system.svc
[ req_dn ]
O = Istio
CN = Intermediate CA
L = east
----
<1> In this example, the configuration file resides in the `east` directory. If your file resides in a different directory, update the syntax accordingly.

> . Create a key for the root certificate by running the following command:

Reviewer: This command creates a certificate signing request, not a key.
```suggestion
. Create a certificate signing request by running the following command:
```

Author: done

+
[source,terminal]
----
  openssl req -new -config east/intermediate.conf \
    -key east/ca-key.pem \
    -out east/cluster-ca.csr
----

> . Create the certificate signing request by running the following command:

Reviewer: This command creates the certificate itself, not the signing request.
```suggestion
. Create the (intermediate CA) certificate (for the east cluster) by running the following command:
```

Author: done

+
[source,terminal]
----  
  openssl x509 -req -sha256 -days 3650 \
    -CA root-cert.pem \
    -CAkey root-key.pem -CAcreateserial \
    -extensions req_ext -extfile east/intermediate.conf \
    -in east/cluster-ca.csr \
    -out east/ca-cert.pem
----

> . Create a shared root certificate for both clusters by running the following command:

Reviewer: This command creates a certificate chain (containing the intermediate and the root CA certificates) for just one cluster, not both.
```suggestion
. Create a certificate chain from the intermediate and root CA certificate for the east cluster:
```

Author: done

+
[source,terminal]
----

>   cat east/ca-cert.pem root-cert.pem \
>     > $east/cert-chain.pem
>   cp root-cert.pem east

Reviewer: ```suggestion
  cat east/ca-cert.pem root-cert.pem > east/cert-chain.pem && cp root-cert.pem east
```

Author: done

----

> 

Reviewer: This is where section 2 (creating the intermediate CA for cluster 1) ends.

> . Create an intermediate certificate for the West cluster by running the following command:

Reviewer: ```suggestion
. Create a key for the intermediate certificate for the West cluster by running the following command:
```

Author: done

+
[source,terminal]
----

> openssl genrsa -out west/ca-key.pem 4096

Reviewer: This also fails because the `west` directory doesn't exist. Insert a step to create the directory (`mkdir west`).

Author: done

----

> . Create an OpenSSL configuration file named `intermediate.conf` for the intermediate certificate of the West cluster by copying the following sample file and saving it locally:

Reviewer: As above, call out the `west/`  directory here.

Author: done

+
.Sample configuration file
[source,subs="attributes,verbatim"]
----

> west/intermediate.conf <1>

Reviewer: Remove this line.

Author: done

[ req ]
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true, pathlen:0
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
subjectAltName=@san
[ san ]
DNS.1 = istiod.istio-system.svc
[ req_dn ]
O = Istio
CN = Intermediate CA
L = west
----
<1> In this example, the configuration file resides in the `west` directory. If your file resides in a different directory, update the syntax accordingly.

> . Create a key for the root certificate by running the following command:

Reviewer: ```suggestion
. Create a certificate signing request by running the following command:
```

Author: done

+
[source,terminal]
----
  openssl req -new -config west/intermediate.conf \
    -key west/ca-key.pem \
    -out west/cluster-ca.csr
----

> . Create the certificate signing request by running the following command:

Reviewer: ```suggestion
. Create the certificate by running the following command:
```

Author: done

+
[source,terminal]
----
  openssl x509 -req -sha256 -days 3650 \
    -CA root-cert.pem \
    -CAkey root-key.pem -CAcreateserial \
    -extensions req_ext -extfile west/intermediate.conf \
    -in west/cluster-ca.csr \
    -out west/ca-cert.pem
----

> . Create the shared root certificate by running the following command:

Reviewer: ```suggestion
. Create the certificate chain by running the following command:
```

Author: done

+
[source,terminal]
----

>   cat west/ca-cert.pem root-cert.pem \
>     > west/cert-chain.pem
>   cp root-cert.pem west

Reviewer: ```suggestion
  cat west/ca-cert.pem root-cert.pem > west/cert-chain.pem && cp root-cert.pem west
```

Author: done

// This module is used in the following assemblies:

// * install/ossm-multi-cluster-topology.adoc

> :_mod-docs-content-type: Procedure

Reviewer 2: ```suggestion
:_mod-docs-content-type: PROCEDURE
```

[id="ossm-creating-certificates-for-a-multi-cluster-mesh_{context}"]
= Creating certificates for a multi-cluster mesh 

Create the root and intermediate certificate authority (CA) certificates for two clusters.

.Prerequisites

* You have OpenSSL installed locally.

.Procedure

> . Create the root CA certificate.

Reviewer 2: ```suggestion
. Create the root CA certificate:
```

Author: done

> .. Create an OpenSSL configuration certificate file named `root-ca.conf` for the root CA certificates:

Reviewer: Flip this step and step "b. Create a key for the root certificate...", so that the order matches the order in "2. Create the intermediate CA certificate for the East cluster" and "3. Create the intermediate CA certificate for the West cluster".

Author: done

+
[source,terminal]
----
$ openssl genrsa -out root-key.pem 4096
----

.. Create an OpenSSL configuration certificate file named `root-ca.conf` for the root CA certificates:
+
.Example root certificate configuration file
[source,terminal]
----
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
[ req_dn ]
O = Istio
CN = Root CA
----

.. Create the certificate signing request by running the following command:
+
[source,terminal]
----
$ openssl req -sha256 -new -key root-key.pem \
  -config root-ca.conf \
  -out root-cert.csr
----

.. Create a shared root certificate by running the following command:
+
[source,terminal]
----
$ openssl x509 -req -sha256 -days 3650 \
  -signkey root-key.pem \
  -extensions req_ext -extfile root-ca.conf \
  -in root-cert.csr \
  -out root-cert.pem
----

> . Create the intermediate CA certificate for the East cluster.

Reviewer 2: ```suggestion
. Create the intermediate CA certificate for the East cluster:
```

Author: done

> .. Create a directory names `east` by running the following command:

Reviewer: ```suggestion
.. Create a directory named `east` by running the following command:
```

Author: done

+
[source,terminal]
----
$ mkdir east
----

.. Create a key for the intermediate certificate for the East cluster by running the following command:
+
[source,terminal]
----
$ openssl genrsa -out east/ca-key.pem 4096
----

.. Create an OpenSSL configuration file named `intermediate.conf` in the `east/` directory for the intermediate certificate of the East cluster. Copy the following example file and save it locally:
+
.Example configuration file
[source,subs="attributes,verbatim"]
----
[ req ]
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true, pathlen:0
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
subjectAltName=@san
[ san ]
DNS.1 = istiod.istio-system.svc
[ req_dn ]
O = Istio
CN = Intermediate CA
L = east
----

.. Create a certificate signing request by running the following command:
+
[source,terminal]
----

> $  openssl req -new -config east/intermediate.conf \

Reviewer 2: Extra space:
```suggestion
$ openssl req -new -config east/intermediate.conf \
```

Author: done

    -key east/ca-key.pem \
    -out east/cluster-ca.csr
----

.. Create the intermediate CA certificate for the East cluster by running the following command:
+
[source,terminal]
----  

> $  openssl x509 -req -sha256 -days 3650 \

Reviewer 2: ```suggestion
$ openssl x509 -req -sha256 -days 3650 \
```

Author: done

    -CA root-cert.pem \
    -CAkey root-key.pem -CAcreateserial \
    -extensions req_ext -extfile east/intermediate.conf \
    -in east/cluster-ca.csr \
    -out east/ca-cert.pem
----

.. Create a certificate chain from the intermediate and root CA certificate for the east cluster by running the following command:
+
[source,terminal]
----
$ cat east/ca-cert.pem root-cert.pem > east/cert-chain.pem && cp root-cert.pem east
----

> . Create the intermediate CA certificate for the West cluster.

Reviewer 2: ```suggestion
. Create the intermediate CA certificate for the West cluster:
```

Author: done

> .. Create a directory names `west` by running the following command:

Reviewer: ```suggestion
.. Create a directory named `west` by running the following command:
```

Author: done. thank you for catching these!

+
[source,terminal]
----
$ mkdir west
----

.. Create a key for the intermediate certificate for the West cluster by running the following command:
+
[source,terminal]
----
$ openssl genrsa -out west/ca-key.pem 4096
----

> .. Create an OpenSSL configuration file named `intermediate.conf` in the `east/` directory for for the intermediate certificate of the West cluster. Copy the following example file and save it locally:

Author: done. thank you!

+
.Example configuration file
[source,subs="attributes,verbatim"]
----
[ req ]
encrypt_key = no
prompt = no
utf8 = yes
default_md = sha256
default_bits = 4096
req_extensions = req_ext
x509_extensions = req_ext
distinguished_name = req_dn
[ req_ext ]
subjectKeyIdentifier = hash
basicConstraints = critical, CA:true, pathlen:0
keyUsage = critical, digitalSignature, nonRepudiation, keyEncipherment, keyCertSign
subjectAltName=@san
[ san ]
DNS.1 = istiod.istio-system.svc
[ req_dn ]
O = Istio
CN = Intermediate CA
L = west
----

.. Create a certificate signing request by running the following command:
+
[source,terminal]
----

> $  openssl req -new -config west/intermediate.conf \

Reviewer 2: ```suggestion
$ openssl req -new -config west/intermediate.conf \
```

Author: done

    -key west/ca-key.pem \
    -out west/cluster-ca.csr
----

.. Create the certificate by running the following command:
+
[source,terminal]
----

> $  openssl x509 -req -sha256 -days 3650 \

Reviewer 2: ```suggestion
$ openssl x509 -req -sha256 -days 3650 \
```

Author: done
